README.md
ali-mts
=======

aliyun mts sdk

## Install

```bash
$ tnpm install ali-mts --save
```

## Usage

```js
var co = require('co');
var oss = require('ali-oss');
var AliyunClient = require('../lib/aliyun_client');

co(function* () {
  var client = new AliyunClient({
    accessKeyId: 'your accessKeyId',
    accessKeySecret: 'your accessKeySecret',
    // set the request timeout,default is 3000
    timeout: 6000
  });

  var result = yield * client.addMedia({
    inputFileUrl: 'http://gxc.oss.aliyuncs.com/ossdemo/VTS_01_1.mp4',
    title: 'VTS_01_1.mp4',
    description: 'xxx',
    tags: 'xxx, yyy, zzz'
  });

  console.log(result);
}).catch(function (err) {
  console.log(err.stack);
});
```

## API




## License

(The MIT License)

Copyright (c) 2014 Alibaba Group Holding Limited.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.1 / 2018-07-06
==================

**others**
,fatal: No names found, cannot describe anything.



lib/oss.js
/**!
 * ali-mts - lib/oss.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var is = require('is-type-of');
var Utils = require('./utils');
var RequestCheckUtils = require('./utils/request_check_utils');

/**
 * 提交转码作业
 *
 * @param {Object} params
 *  - input  作业输入,JSON对象
 *  - outputBucket  输出Bucket,需在MTS控制台中资源控制频道里的Bucket授权页面授予此Bucket写权限给MTS
 *  - outputLocation  输出 Bucket 所在数据中心,默认值是oss-cn-hangzhou
 *  - outputs  Outputs由Output列表构成,JSON数组,大小上限为10
 */
exports.submitJobs = function* (params) {
  params = params || {};
  if (is.object(params.input)) {
    params.input = JSON.stringify(Utils.normalizeObject(params.input));
  }
  if (is.object(params.outputs)) {
    params.outputs = JSON.stringify(Utils.normalizeObject(params.outputs));
  }
  params.outputLocation = params.outputLocation || 'oss-cn-hangzhou';

  RequestCheckUtils.checkNotEmpty(params.input, 'input');
  RequestCheckUtils.checkNotEmpty(params.outputBucket, 'outputBucket');
  RequestCheckUtils.checkNotEmpty(params.outputs, 'outputs');
  RequestCheckUtils.checkNotEmpty(params.pipelineId, 'pipelineId');

  var result = yield * this.invoke('mts.aliyuncs.com.SubmitJobs.2014-06-18', params);
  return result;
};

/**
 * 取消转码作业
 *
 * @param {Object} params
 *  - jobId  作业ID
 */
exports.cancelJob = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.jobId, 'jobId');

  var result = yield * this.invoke('mts.aliyuncs.com.CancelJob.2014-06-18', params);
  return result;
};

/**
 * 查询转码作业
 *
 * @param {Object} params
 *  - jobIds  作业ID列表,逗号分隔,一次最多10个
 */
exports.queryJobList = function* (params) {
  params = params || {};

  var result = yield * this.invoke('mts.aliyuncs.com.QueryJobList.2014-06-18', params);
  return result;
};

/**
 * 搜索转码作业
 *
 * @param {Object} params
 *  - pageNumber  当前页号,从1开始,默认值是1
 *  - pageSize  分页查询时设置的每页大小,默认值是10,最大不超过100
 *  - state  转码任务状态:All表示所有状态,Submitted表示作业已提交,Transcoding表示转码 中,TranscodeSuccess表示转码成功,TranscodeFail表示转码失败,TranscodeCancelled表示 转码取消,默认是All
 *  - startOfJobCreatedTimeRange  创建转码作业时间范围中的下限值,日期格式按照ISO8601标准表示,并需要使用UTC时间
 *  - endOfJobCreatedTimeRange  创建转码作业时间的上限,日期格式按照ISO8601标准表示,并需要使用UTC时间
 */
exports.searchJob = function* (params) {
  params = params || {};
  if (is.date(params.startOfJobCreatedTimeRange)) {
    params.startOfJobCreatedTimeRange =
      Utils.formatDate(params.startOfJobCreatedTimeRange, 'UTC:yyyy-mm-dd\'T\'HH:MM:ss\'Z\'');
  }
  if (is.date(params.endOfJobCreatedTimeRange)) {
    params.endOfJobCreatedTimeRange =
      Utils.formatDate(params.endOfJobCreatedTimeRange, 'UTC:yyyy-mm-dd\'T\'HH:MM:ss\'Z\'');
  }

  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchJob.2014-06-18', params);
  return result;
};

/**
 * 通过管道查询转码作业
 *
 * @param {Object} params
 *  - pipelineId  管道ID
 *  - pageNumber  当前页号,从1开始,默认值是1
 *  - pageSize  分页查询时设置的每页大小,默认值是10,最大不超过100
 */
exports.queryJobListByPid = function* (params) {
  params = params || {};

  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryJobListByPid.2014-06-18', params);
  return result;
};

lib/pipeline.js
/**!
 * ali-mts - lib/media.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var RequestCheckUtils = require('./utils/request_check_utils');

/**
 * 搜索管道
 *
 * @param {Object} params
 *  - pageNumber  当前页号,从第1页开始,默认值是1
 *  - pageSize  分页查询时设置的每页行数页面大小,默认值是10,上限100
 *  - state 管道状态:All、Active、Paused、Deleted,默认是All
 */
exports.searchPipeline = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchPipeline.2014-06-18', params);
  return result;
};

/**
 * 查询管道
 *
 * @param {Object} params
 *  - pipelineIds  管道ID列表,最多一次查10个,逗号分隔
 */
exports.queryPipelineList = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.pipelineIds, 'pipelineIds');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryPipelineList.2014-06-18', params);
  return result;
};

/**
 * 更新管道
 *
 * @param {Object} params
 *  - pipelineId  管道ID
 *  - name  管道名称,最大长度128字节
 *  - state 管道状态,分为Active、Paused两种
 */
exports.updatePipeline = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxLength(params.name, 128, 'name');
  RequestCheckUtils.checkNotEmpty(params.pipelineId, 'pipelineId');

  var result = yield * this.invoke('mts.aliyuncs.com.UpdatePipeline.2014-06-18', params);
  return result;
};

lib/template.js
/**!
 * ali-mts - lib/template.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var is = require('is-type-of');
var Utils = require('./utils');
var RequestCheckUtils = require('./utils/request_check_utils');

/**
 * 新增自定义模板
 *
 * @param {Object} params
 *  - name 模板名称,最大长度128字节
 *  - container 容器,JSON对象
 *  - video 视频流配置,JSON对象
 *  - audio 音频流配置,JSON对象
 */
exports.addTemplate = function* (params) {
  params = params || {};
  if (is.object(params.container)) {
    params.container = JSON.stringify(Utils.normalizeObject(params.container));
  }
  if (is.object(params.video)) {
    params.video = JSON.stringify(Utils.normalizeObject(params.video));
  }
  if (is.object(params.audio)) {
    params.audio = JSON.stringify(Utils.normalizeObject(params.audio));
  }

  RequestCheckUtils.checkNotEmpty(params.name, 'name');
  RequestCheckUtils.checkMaxLength(params.name, 128, 'name');

  var result = yield * this.invoke('mts.aliyuncs.com.AddTemplate.2014-06-18', params);
  return result;
};

/**
 * 搜索自定义模板
 *
 * @param {Object} params
 *  - name 模板名称,最大长度128字节
 *  - pageNumber 当前页号,从第1页开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 *  - state 转码模板状态:All表示所有,Normal表示正常,Deleted表示已删除,默认是All
 */
exports.searchTemplate = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchTemplate.2014-06-18', params);
  return result;
};

/**
 * 查询自定义模板
 *
 * @param {Object} params
 *  - templateIds 模板ID列表,最多一次查10个,逗号分隔
 */
exports.queryTemplateList = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.templateIds, 'templateIds');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryTemplateList.2014-06-18', params);
  return result;
};

/**
 * 更新自定义转码模板
 *
 * @param {Object} params
 *  - templateId 模板ID
 *  - name 模板名称,最大长度128字节
 *  - container 容器,JSON对象
 *  - video 视频流配置,JSON对象
 *  - audio 音频流配置,JSON对象
 */
exports.updateTemplate = function* (params) {
  params = params || {};
  if (is.object(params.container)) {
    params.container = JSON.stringify(Utils.normalizeObject(params.container));
  }
  if (is.object(params.video)) {
    params.video = JSON.stringify(Utils.normalizeObject(params.video));
  }
  if (is.object(params.audio)) {
    params.audio = JSON.stringify(Utils.normalizeObject(params.audio));
  }

  RequestCheckUtils.checkMaxLength(params.name, 128, 'name');
  RequestCheckUtils.checkNotEmpty(params.templateId, 'templateId');

  var result = yield * this.invoke('mts.aliyuncs.com.UpdateTemplate.2014-06-18', params);
  return result;
};

/**
 * 删除自定义转码模板
 *
 * @param {Object} params
 *  - templateId 模板ID
 */
exports.deleteTemplate = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.templateId, 'templateId');

  var result = yield * this.invoke('mts.aliyuncs.com.DeleteTemplate.2014-06-18', params);
  return result;
};

lib/watermark.js
/**!
 * ali-mts - lib/watermark.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var is = require('is-type-of');
var Utils = require('./utils');
var RequestCheckUtils = require('./utils/request_check_utils');


/**
 * 新增水印模板
 *
 * @param {Object} params
 *  - name 模板名称,最大长度128字节
 *  - config 水印模板配置,JSON对象
 */
exports.addWaterMarkTemplate = function* (params) {
  params = params || {};
  if (is.object(params.config)) {
    params.config = JSON.stringify(Utils.normalizeObject(params.config));
  }

  RequestCheckUtils.checkNotEmpty(params.config, 'config');
  RequestCheckUtils.checkNotEmpty(params.name, 'name');
  RequestCheckUtils.checkMaxLength(params.name, 128, 'name');

  var result = yield * this.invoke('mts.aliyuncs.com.AddWaterMarkTemplate.2014-06-18', params);
  return result;
};

/**
 * 搜索水印模板
 *
 * @param {Object} params
 *  - pageNumber 当前页号,从第1页开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 *  - state 水印模板状态:All表示所有,Normal表示正常,Deleted表示已删除的,默认是All
 */
exports.searchWaterMarkTemplate = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchWaterMarkTemplate.2014-06-18', params);
  return result;
};

/**
 * 查询水印模板
 *
 * @param {Object} params
 *  - pageNumber 当前页号,从第1页开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 *  - state 水印模板状态:All表示所有,Normal表示正常,Deleted表示已删除的,默认是All
 */
exports.queryWaterMarkTemplateList = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.waterMarkTemplateIds, 'waterMarkTemplateIds');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryWaterMarkTemplateList.2014-06-18', params);
  return result;
};

/**
 * 删除水印模板
 *
 * @param {Object} params
 *  - waterMarkTemplateId 水印模板ID
 */
exports.deleteWaterMarkTemplate = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.waterMarkTemplateId, 'waterMarkTemplateId');

  var result = yield * this.invoke('mts.aliyuncs.com.DeleteWaterMarkTemplate.2014-06-18', params);
  return result;
};

lib/utils/index.js
/**!
 * ali-mts - lib/aliyun_client.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var debug = require('debug')('aliyun:mts');
var crypto = require('crypto');
var is = require('is-type-of');

/**
 * 首字母大写
 */
exports.ucFirst = function (val) {
  if (!val) return val;

  return val.slice(0, 1).toUpperCase() + val.slice(1);
};

/**
 * 将对象格式化成首字母大写形式
 */
exports.normalizeObject = function (obj) {
  if (!is.object(obj)) return obj;

  var newObj = {};
  for (var key in obj) {
    if (obj.hasOwnProperty(key)) {
      newObj[ucFirst(key)] = obj[key];
    }
  }
  return newObj;
};


/**
 *  aliyun sdk 签名方法
 *
 *  MTS会对每个访问的请求进行身份验证,所以无论使用HTTP还是HTTPS协议提交请求,
 *  都需要在请求中包含签名(Signature)信息。MTS通过使用Access Key ID 和Access Key Secret
 *  进行对称加密的方法来验证请求的发送者身份。Access Key ID和Access Key Secret由阿里云官
 *  方颁发给访问者(可以通过阿里云官方网站申 请和管理),其中Access Key ID用于标识访问者身份;
 *  Access Key Secret是用于加密签名字符串和服务器端验证签名字符串的密钥,必须严格保密,只有
 *  阿里云和 用户知道。
 *
 * @param {Object} parameters
 * @param {String} accessKeySecret
 */
exports.computeSignature = function (parameters, accessKeySecret) {
  parameters = parameters || {};

  // 将参数Key按字典顺序排序
  var sortedKeys = Object.keys(parameters).sort();
  var canonicalizedQueryString = '';
  sortedKeys.forEach(function (key) {
    if (!parameters[key]) return;

    canonicalizedQueryString += ('&' + encodeURIComponent(key) + '=' + encodeURIComponent(parameters[key]));
  });
  var stringToSign = 'POST&' + encodeURIComponent('/') +
    '&' + encodeURIComponent(canonicalizedQueryString.slice(1));
  debug('stringToSign: %s', stringToSign);

  // 计算签名时使用的Key就是用户持有的AccessKeySecret并加上一个"&"字 符(ASCII:38),使用的哈希算法是SHA1。
  var signature = crypto.createHmac('sha1', accessKeySecret + '&');
  debug('signature: %s', signature);

  return signature.update(stringToSign).digest('base64');
};

/**
 * 将参数转换成 url querystring
 */
exports.paramsToQueryString = function (params) {
  if (params == null) {
    return '';
  }

  var paramString = '';
  var isFirst = true;

  Object.keys(params).forEach(function (key) {
    if (params[key] == null) return;

    if (isFirst) {
      paramString += encode(key);
      isFirst = false;
    } else {
      paramString += ('&' + encode(key));
    }
    if (params[key]) {
      paramString += ('=' + encode(params[key]));
    }
  });

  return paramString;
};

// todo: confirm
function encode(val) {
  return encodeURIComponent(val)
    .replace(/%20/ig, '+')
    .replace(/%2A/ig, '*')
    .replace(/~/ig, '%7E');
}


// Some common format strings
var masks = {
  "default": "ddd mmm dd yyyy HH:MM:ss",
  shortDate: "m/d/yy",
  mediumDate: "mmm d, yyyy",
  longDate: "mmmm d, yyyy",
  fullDate: "dddd, mmmm d, yyyy",
  shortTime: "h:MM TT",
  mediumTime: "h:MM:ss TT",
  longTime: "h:MM:ss TT Z",
  isoDate: "yyyy-mm-dd",
  isoTime: "HH:MM:ss",
  isoDateTime: "yyyy-mm-dd'T'HH:MM:ss",
  isoUtcDateTime: "UTC:yyyy-mm-dd'T'HH:MM:ss'Z'"
};

// Internationalization strings
var i18n = {
  dayNames: [
    "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat",
    "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
  ],
  monthNames: [
    "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
    "January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"
  ]
};

var token = /d{1,4}|m{1,4}|yy(?:yy)?|([HhMsTt])\1?|[LloSZ]|"[^"]*"|'[^']*'/g;
var timezone = /\b(?:[PMCEA][SDP]T|(?:Pacific|Mountain|Central|Eastern|Atlantic) (?:Standard|Daylight|Prevailing) Time|(?:GMT|UTC)(?:[-+]\d{4})?)\b/g;
var timezoneClip = /[^-+\dA-Z]/g;
var pad = function (val, len) {
  val = String(val);
  len = len || 2;
  while (val.length < len) val = "0" + val;
  return val;
};

// Regexes and supporting functions are cached through closure
exports.formatDate = function (date, mask, utc) {
  // You can't provide utc if you skip other args (use the "UTC:" mask prefix)
  if (arguments.length == 1 && Object.prototype.toString.call(date) == "[object String]" && !/\d/.test(date)) {
    mask = date;
    date = undefined;
  }

  // Passing date through Date applies Date.parse, if necessary
  date = date ? new Date(date) : new Date;
  if (isNaN(date)) throw SyntaxError("invalid date");

  mask = String(masks[mask] || mask || masks["default"]);

  // Allow setting the utc argument via the mask
  if (mask.slice(0, 4) == "UTC:") {
    mask = mask.slice(4);
    utc = true;
  }

  var _ = utc ? "getUTC" : "get",
    d = date[_ + "Date"](),
    D = date[_ + "Day"](),
    m = date[_ + "Month"](),
    y = date[_ + "FullYear"](),
    H = date[_ + "Hours"](),
    M = date[_ + "Minutes"](),
    s = date[_ + "Seconds"](),
    L = date[_ + "Milliseconds"](),
    o = utc ? 0 : date.getTimezoneOffset(),
    flags = {
      d: d,
      dd: pad(d),
      ddd: i18n.dayNames[D],
      dddd: i18n.dayNames[D + 7],
      m: m + 1,
      mm: pad(m + 1),
      mmm: i18n.monthNames[m],
      mmmm: i18n.monthNames[m + 12],
      yy: String(y).slice(2),
      yyyy: y,
      h: H % 12 || 12,
      hh: pad(H % 12 || 12),
      H: H,
      HH: pad(H),
      M: M,
      MM: pad(M),
      s: s,
      ss: pad(s),
      l: pad(L, 3),
      L: pad(L > 99 ? Math.round(L / 10) : L),
      t: H < 12 ? "a" : "p",
      tt: H < 12 ? "am" : "pm",
      T: H < 12 ? "A" : "P",
      TT: H < 12 ? "AM" : "PM",
      Z: utc ? "UTC" : (String(date).match(timezone) || [""]).pop().replace(timezoneClip, ""),
      o: (o > 0 ? "-" : "+") + pad(Math.floor(Math.abs(o) / 60) * 100 + Math.abs(o) % 60, 4),
      S: ["th", "st", "nd", "rd"][d % 10 > 3 ? 0 : (d % 100 - d % 10 != 10) * d % 10]
    };

  return mask.replace(token, function ($0) {
    return $0 in flags ? flags[$0] : $0.slice(1, $0.length - 1);
  });
};

lib/utils/request_check_utils.js
/**!
 * ali-mts - lib/utils/request_check_utils.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var ERROR_CODE_ARGUMENTS_MISS = '40';
var ERROR_CODE_ARGUMENTS_INVALID = '41';

exports.checkNotEmpty = function (val, fieldName) {
  if (!val) {
    var err = new Error('client-error:Missing Required Arguments:' + fieldName);
    err.code = ERROR_CODE_ARGUMENTS_MISS;
    throw err;
  }
};

exports.checkMaxLength = function (val, maxLength, fieldName) {
  if (val == null) return;

  if (val.length > maxLength) {
    var err = new Error('client-error:Invalid Arguments:the length of ' + fieldName + ' can not be larger than ' + maxLength + '.');
    err.code = ERROR_CODE_ARGUMENTS_INVALID;
    throw err;
  }
};

exports.checkMaxValue = function (val, maxValue, fieldName) {
  if (val == null) return;

  if (val > maxValue) {
    var err = new Error('client-error:Invalid Arguments:the value of ' + fieldName + ' can not be larger than ' + maxValue + '.');
    err.code = ERROR_CODE_ARGUMENTS_INVALID;
    throw err;
  }
};

exports.checkMinValue = function (val, minValue, fieldName) {
  if (val == null) return;

  if (val < minValue) {
    var err = new Error('client-error:Invalid Arguments:the value of ' + fieldName + ' can not be smaller than ' + minValue + '.');
    err.code = ERROR_CODE_ARGUMENTS_INVALID;
    throw err;
  }
};

lib/media.js
/**!
 * ali-mts - lib/media.js
 *
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 * Authors:
 *   宗羽 <xiaochen.gaoxc@alibaba-inc.com> (http://github.com/gxcsoccer)
 */

'use strict';

/**
 * Module dependencies.
 */

var is = require('is-type-of');
var Utils = require('./utils');
var RequestCheckUtils = require('./utils/request_check_utils');

/**
 * 添加媒资
 *
 * @param {Object} params
 *  - inputFileUrl  输入文件的OSS 地址,地址必须符合OSS Object 外链地址规则
 *  - title 媒资的标题
 *  - tags  资的标签,多个标签用逗号分隔
 *  - description 媒资的描述
 */
exports.addMedia = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxLength(params.description, 512, 'description');
  RequestCheckUtils.checkNotEmpty(params.inputFileUrl, 'inputFileUrl');
  RequestCheckUtils.checkMaxLength(params.inputFileUrl, 1152, 'inputFileUrl');
  RequestCheckUtils.checkMaxLength(params.tags, 256, 'tags');
  RequestCheckUtils.checkNotEmpty(params.title, 'title');
  RequestCheckUtils.checkMaxLength(params.title, 256, 'title');

  var result = yield * this.invoke('mts.aliyuncs.com.AddMedia.2014-06-18', params);
  return result;
};

/**
 * 查询媒资
 *
 * @param {Object} params
 *  - mediaIds  媒资ID集合,ID之间采用逗号分隔,一次最多传入10个ID
 */
exports.queryMediaList = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaIds, 'mediaIds');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaList.2014-06-18', params);
  return result;
};

/**
 * 搜索媒资
 *
 * @param {Object} params
 *  - pageNumber  当前页码,起始值为 1,不指定时,默认值为 1
 *  - pageSize  分页查询时设置的每页行数,不指定时,默认值为 10,指定时,最大不超过100
 *  - state 媒资状态:All表示所有,Normal表示正常,Deleted表示已删除的,默认是All
 */
exports.searchMedia = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchMedia.2014-06-18', params);
  return result;
};

/**
 * 更新媒资
 *
 * @param {Object} params
 *  - mediaId  媒资ID
 *  - title  媒资的标题,最大长度256字节
 *  - tags  媒资的标签,多个标签用空格分隔
 *  - coverUrl  媒资的封面图片OSS地址,符合OSS Object外链规则
 *  - description 媒资的描述
 */
exports.updateMedia = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxLength(params.coverUrl, 1152, 'coverUrl');
  RequestCheckUtils.checkMaxLength(params.description, 512, 'description');
  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');
  RequestCheckUtils.checkMaxLength(params.tags, 256, 'tags');
  RequestCheckUtils.checkMaxLength(params.title, 256, 'title');

  var result = yield * this.invoke('mts.aliyuncs.com.UpdateMedia.2014-06-18', params);
  return result;
};

/**
 * 删除媒资
 *
 * @param {Object} params
 *  - mediaId  媒资ID
 */
exports.deleteMedia = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');

  var result = yield * this.invoke('mts.aliyuncs.com.DeleteMedia.2014-06-18', params);
  return result;
};

/**
 * 查询媒资可用预置模版
 *
 * @param {Object} params
 *  - mediaId  媒资ID
 */
exports.queryMediaAnalysis = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaAnalysis.2014-06-18', params);
  return result;
};

/**
 * 提交媒资作业
 *
 * @param {Object} params
 *  - mediaId  媒资ID
 *  - outputs Outputs由Output列表构成,JSON数组,大小上限为10
 *  - pipelineId  管道ID
 */
exports.submitMediaJobs = function* (params) {
  params = params || {};
  if (is.object(params.outputs)) {
    params.outputs = JSON.stringify(Utils.normalizeObject(params.outputs));
  }

  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');
  RequestCheckUtils.checkNotEmpty(params.outputs, 'outputs');

  var result = yield * this.invoke('mts.aliyuncs.com.SubmitMediaJobs.2014-06-18', params);
  return result;
};

/**
 * 查询媒资作业
 *
 * @param {Object} params
 *  - mediaJobIds 媒资作业ID列表,逗号分隔,上限10个
 */
exports.queryMediaJobList = function* (params) {
  params = params || {};

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaJobList.2014-06-18', params);
  return result;
};

/**
 * 取消媒资作业
 *
 * @param {Object} params
 *  - mediaJobId 媒资作业ID
 */
exports.cancelMediaJob = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaJobId, 'mediaJobId');

  var result = yield * this.invoke('mts.aliyuncs.com.CancelMediaJob.2014-06-18', params);
  return result;
};

/**
 * 搜索媒资作业
 *
 * @param {Object} params
 *  - pageNumber 当前页号,从1开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 *  - state 媒资作业状态:All表示所有状态,Submitted表示作业已提交,Transcoding表示转码 中,TranscodeSuccess表示转码成功,TranscodeFail表示转码失败,TranscodeCancelled表示 转码取消,默认是All
 *  - startOfJobCreatedTimeRange 创建媒资作业时间范围中的下限值,日期格式按照ISO8601标准表示,并需要使用UTC时间。格 式为:YYYY-MM-DDThh:mm:ssZ 例如,2014-01-10T12:00:00Z(为北京时间2014年1月10 日20点0分0秒)
 *  - endOfJobCreatedTimeRange 创建媒资作业时间的上限,日期格式按照ISO8601标准表示,并需要使用UTC时间。格式为: YYYY-MM-DDThh:mm:ssZ 例如,2014-01-11T12:00:00Z(为北京时间2014年1月11日20点 0分0秒)
 */
exports.searchMediaJob = function* (params) {
  params = params || {};
  if (is.date(params.startOfJobCreatedTimeRange)) {
    params.startOfJobCreatedTimeRange =
      Utils.formatDate(params.startOfJobCreatedTimeRange, 'UTC:yyyy-mm-dd\'T\'HH:MM:ss\'Z\'');
  }
  if (is.date(params.endOfJobCreatedTimeRange)) {
    params.endOfJobCreatedTimeRange =
      Utils.formatDate(params.endOfJobCreatedTimeRange, 'UTC:yyyy-mm-dd\'T\'HH:MM:ss\'Z\'');
  }

  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.SearchMediaJob.2014-06-18', params);
  return result;
};

/**
 * 通过媒资查询媒资作业
 *
 * @param {Object} params
 *  - mediaIds 媒资ID,每次最多只能查十个,以逗号分隔
 *  - state 媒资作业状态:All表示所有状态,Submitted表示作业已提交,Transcoding表示转码 中,TranscodeSuccess表示转码成功,TranscodeFail表示转码失败,TranscodeCancelled表示 转码取消,默认是All
 */
exports.queryMediaJobListByMids = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaIds, 'mediaIds');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaJobListByMids.2014-06-18', params);
  return result;
};

/**
 * 通过管道查询媒资作业
 *
 * @param {Object} params
 *  - pipelineId 管道ID
 *  - pageNumber 当前页号,从第1页开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 */
exports.queryMediaJobListByPid = function* (params) {
  params = params || {};
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaJobListByPid.2014-06-18', params);
  return result;
};

/**
 * 截取媒资图像
 *
 * @param {Object} params
 *  - mediaId 媒资ID
 *  - time 截取的时间点,单位毫秒
 */
exports.mediaSnapshot = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');
  RequestCheckUtils.checkNotEmpty(params.time, 'time');

  var result = yield * this.invoke('mts.aliyuncs.com.MediaSnapshot.2014-06-18', params);
  return result;
};

/**
 * 查询媒资图像
 *
 * @param {Object} params
 *  - mediaId 媒资ID
 *  - pageNumber 当前页号,从第1页开始,默认值是1
 *  - pageSize 分页查询时设置的每页大小,默认值是10,上限100
 */
exports.queryMediaSnapshotList = function* (params) {
  params = params || {};
  RequestCheckUtils.checkNotEmpty(params.mediaId, 'mediaId');
  RequestCheckUtils.checkMaxValue(params.pageSize, 100, 'pageSize');
  RequestCheckUtils.checkMinValue(params.pageSize, 1, 'pageSize');

  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaSnapshotList.2014-06-18', params);
  return result;
};

/**
 * 查询媒资Bucket
 */
exports.queryMediaBucket = function* () {
  var result = yield * this.invoke('mts.aliyuncs.com.QueryMediaBucket.2014-06-18');
  return result;
}